docs/architecture/adr-025-ibc-passive-channels.md

```diff
@@ -0,0 +1,126 @@
+      sdk.NewAttribute(types.AttributeKeyOrder, order.String()),
+      sdk.NewAttribute(types.AttributeKeySrcPort, portID),
+      sdk.NewAttribute(types.AttributeKeySrcChannel, chanenlID),
+      sdk.NewAttribute(types.AttributeKeySrcVersion, version),
+      sdk.NewAttribute(types.AttributeKeyDstPort, counterparty.GetPortID()),
+      sdk.NewAttribute(types.AttributeKeyDstChannel, counterparty.GetChannelID()),
+      // The destination version is not yet known, but a value is necessary to pad
+      // the event attribute offsets
+      sdk.NewAttribute(types.AttributeKeyDstVersion, ""),
+    ),
+  })
+```
+
+These metadata events capture all the "header" information needed to route IBC channel handshake transactions without requiring the client to query any data except that of the connection ID that it is willing to relay.  It is intended that `channel_meta.src_connection` is the only event key that needs to be indexed for a passive relayer to function.
+
+### Handling Channel Open Attempts
+
+In the case of the passive relayer, when one chain sends a `ChanOpenInit`, the relayer should inform the other chain of this open attempt and allow that chain to decide how (and if) it continues the handshake.  Once both chains have actively approved the channel opening, then the rest of the handshake can happen as it does with the current "naive" relayer.
+
+To implement this behavior, we propose adding a new callback `cbs.OnAttemptChanOpenTry` which explicitly handles the `MsgChannelOpenTry`, usually by resulting in a call to `keeper.ChanOpenTry`.  If the callback is not supplied, then the default behaviour would be to use `channel.HandleMsgChannelOpenTry`, for compatibility with existing chains that expect a "naive" relayer.
+
+Here is how this callback would be used, in the implementation of `x/ibc/handler.go`:
+
+```go
+// Declare an interface for handling a ChanOpenTry.
+type AttemptChanOpenTryCallback interface {
+  OnAttemptChanOpenTry(ctx sdk.Context, k keeper.Keeper, portCap *capability.Capability, msg types.MsgChannelOpenTry) (*sdk.Result, error)
+}
+// ...
+    case channel.MsgChannelOpenTry:
+      // Lookup module by port capability
+      module, portCap, err := k.PortKeeper.LookupModuleByPort(ctx, msg.PortID)
+      if err != nil {
+              return nil, sdkerrors.Wrap(err, "could not retrieve module from port-id")
+      }
+      // =======================================
+      // NEW CODE: Check if the module defines an HandleMsgChannelOpenTry callback.
+      // Retrieve callbacks from router
+      cbs, ok := k.Router.GetRoute(module)
+      if !ok {
+              return nil, sdkerrors.Wrapf(port.ErrInvalidRoute, "route not found to module: %s", module)
+      }
+      if tryHandler, ok := cbs.(AttemptChanOpenTryCallback); ok {
+        // Allow the port's try handler to override the default OpenTry behaviour.
+        return tryHandler.OnAttemptChanOpenTry(ctx, k.ChannelKeeper, portCap, msg)
```

Comment:
I don’t understand what make this function different from the default handler below

Reply:
Instead of using predetermined `HandleMsgChannelOpenTry` behaviour (which unconditionally continues the handshake with the version string that was supplied by the `MsgChannelOpenTry`), this gives the port's app module a chance to do something different.  Notably, it might choose not to continue the handshake immediately, and/or to choose a different version string.

Reply:
Maybe it is actually cleaner to always require that the module define `OnAttemptChanOpenTry`, and provide a default implementation for that which calls `channel.HandleMsgChannelOpenTry` - what do you think? I generally prefer to avoid a lot of "alternative call paths" in favour of defaults if possible.

Reply:
> Maybe it is actually cleaner to always require that the module define `OnAttemptChanOpenTry`, and provide a default implementation for that which calls `channel.HandleMsgChannelOpenTry` - what do you think?

I like this idea.  How would I "provide a default implementation" in practice?  Just by example in the `ibc-transfer` code, or is there some other mechanism you have in mind?

Reply:
Yup, in `module.go` probably.

Reply:
Done.